### PR TITLE
Add vectorized Jacobian and Hessian computation with forward AD

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68629
* #68628
* #67045
* __->__ #67041
* #67820
* #66294
* #66293
* #66292

Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -6781,10 +6781,15 @@ def test_jacobian_create_graph(self):
...
+    def _check_jacobian_vectorize_correctness(self, f, inputs, test_forward_ad=True):
         expected = autogradF.jacobian(f, inputs, vectorize=False)
-        result = autogradF.jacobian(f, inputs, vectorize=True)
-        self.assertEqual(result, expected)
+        result_backward_mode = autogradF.jacobian(f, inputs, vectorize=True)
+        self.assertEqual(result_backward_mode, expected)
+
+        if test_forward_ad:
+            result_forward_mode = autogradF.jacobian(f, inputs, forward_ad=True, vectorize=True)
+            print(result_forward_mode, expected)
```
Comment (albanD) [inline test/test_autograd.py:6791]: debug print

File: test/test_autograd.py [hunk 2]
```diff
@@ -6856,13 +6861,18 @@ def f(x, y):
 
         x = torch.randn(3)
         y = torch.randn(3)
-        self._check_jacobian_vectorize_correctness(f, (x, y))
+        # The Jacobin computed using forward AD has the dtype of the output
```
Comment (albanD) [inline test/test_autograd.py:6864]: Is that a problem? Should we document that in the jacobian function?

File: torch/autograd/functional.py [hunk 3]
```diff
@@ -408,7 +421,66 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
...
+                           'Please either set `strict=False` or '
+                           '`forward_ad=False`.')
+    is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "jacobian")
+    outputs = func(*inputs)
+    is_outputs_tuple, outputs = _as_tuple(outputs,
+                                          "outputs of the user-provided function",
+                                          "jacobian")
+    if vectorize:
+        # See NOTE: [Computing jacobian with vmap and grad for multiple outputs]
+        if strict:
```
Comment (albanD) [inline torch/autograd/functional.py:437]: This is already checked above.

File: torch/autograd/functional.py [hunk 4]
```diff
@@ -408,7 +421,66 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
...
 
-def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
+def _jacfwd(func, inputs, strict=False, vectorize=False):
+    if strict:
+        raise RuntimeError('torch.autograd.functional.jacobian: `strict=True` '
+                           'and `forward_ad=True` are not supported together (yet). '
+                           'Please either set `strict=False` or '
+                           '`forward_ad=False`.')
+    is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "jacobian")
+    outputs = func(*inputs)
```
Comment (albanD) [inline torch/autograd/functional.py:431]: Why do you need this? Your jvp function below can also return the output no?

File: torch/autograd/functional.py [hunk 5]
```diff
@@ -472,6 +544,13 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
          tensor([[3., 0.],
                  [0., 3.]]))
     """
+    if forward_ad:
+        if create_graph:
+            raise RuntimeError('torch.autograd.functional.jacobian: `create_graph=True` '
+                               'and `forward_ad=True` are not supported together (yet). '
+                               'Please either set `strict=False` or '
```
Comment (albanD) [inline torch/autograd/functional.py:551]: That's not the right recommendation?

File: torch/autograd/functional.py [hunk 6]
```diff
@@ -472,6 +544,13 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
          tensor([[3., 0.],
                  [0., 3.]]))
     """
+    if forward_ad:
+        if create_graph:
+            raise RuntimeError('torch.autograd.functional.jacobian: `create_graph=True` '
```
Comment (albanD) [inline torch/autograd/functional.py:549]: Why is this one not supported? Just because we don't test it?
Also should this be a NotImplementedError ?

File: torch/autograd/functional.py [hunk 7]
```diff
@@ -690,11 +769,14 @@ def ensure_single_output_function(*inp):
         return out.squeeze()
 
     def jac_func(*inp):
+        if forward_ad:
+            # Ensure the graph is created (and not detached)
```
Comment (albanD) [inline torch/autograd/functional.py:773]: What does this mean? Is that for the create_graph case?

File: test/test_autograd.py (hunk 2 above)
Comment (soulitzer) [reply test/test_autograd.py:6864]: Not sure if it is a problem. But yeah, it's probably good to document.

File: torch/autograd/functional.py (hunk 6 above)
Comment (soulitzer) [reply torch/autograd/functional.py:549]: Yeah I haven't tested it. Just putting NotImplementedError for now

File: torch/autograd/functional.py (hunk 7 above)
Comment (soulitzer) [reply torch/autograd/functional.py:773]: Currently for  in this wrapper make sure inputs have requires_grad=True, and still create_graph=True so that input's won't be detached in jacobian's preprocessing (which detaches when either create_graph=False or requires_grad of its input is False). This way avoids changing jacobian's API but has more overhead because we must create the graph.

Alternatively we could add a for_forward_ad_hessian flag which will skip the _grad_preprocess stage and avoid detaching the inputs while also avoiding the creation of the backward graph

File: torch/autograd/functional.py (hunk 4 above)
Comment (soulitzer) [reply torch/autograd/functional.py]: There needed to be some more logic to extract the output from the wrapped vmap(dual, but it is fixed now

File: torch/autograd/functional.py (hunk 7 above)
Comment (albanD) [reply torch/autograd/functional.py]: Ho this was done before the idea that detach applies both to forward and backward mode AD

Should we replace the `.detach()` by a 
```
with torch.no_grad():
    inp = inp.clone()
```
when forward AD is used?
I guess not and we should introduce a new API to detach backward grads only <- TODO for forward AD project

**Review by albanD — APPROVED** [review]
some lint failures but lgtm otherwise

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

File: torch/autograd/functional.py [hunk 8]
```diff
@@ -598,7 +679,7 @@ def vjp(grad_output):
 
         return _tuple_postprocess(jacobian, (is_outputs_tuple, is_inputs_tuple))
 
-def hessian(func, inputs, create_graph=False, strict=False, vectorize=False):
+def hessian(func, inputs, create_graph=False, strict=False, vectorize=False, use_forward_ad=False):
```
Comment (zou3519) [inline torch/autograd/functional.py:682]: You probably want to name this something more descriptive. If I see `hessian(..., use_forward_ad=True)`, then I might assume that we're using forward-mode AD for both jacobian computations. Maybe "use_forward_ad_for_outer_jacobian"? Or "outer_jacobian_strategy = {forward, reverse}"

File: torch/autograd/functional.py [hunk 9]
```diff
@@ -408,7 +421,64 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
...
+            jacobian_input_output.append(jacobian_output_i_output)
+
+        # Omit [Step 4] because everything is already transposed w/ forward AD
+        return _tuple_postprocess(jacobian_input_output, (is_outputs_tuple, is_inputs_tuple))
+    else:
+        raise NotImplementedError("Computing Jacobian using forward-AD or forward-over-reverse Hessian is"
+                                  "only implemented for `vectorize=True`.")
+
+
+def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False, use_forward_ad=False):
```
Comment (zou3519) [inline torch/autograd/functional.py:481]: `use_forward_ad` should probably be more neutral, like "jacobian(strategy={"forward-mode", "reverse-mode"})"

**Comment by zou3519** [comment]
By the way, I'm not sure that "strategy" is the right kwarg (it was the first thing that came to mind), should we try to bikeshed it some more?

**Comment by soulitzer** [comment]
> By the way, I'm not sure that "strategy" is the right kwarg (it was the first thing that came to mind), should we try to bikeshed it some more?

I guess more generic alternatives could be: `"mode"`, `"method"`, or `"algorithm"`. Doing a quick search of how we name str fields `(str):` or `(str, optional):`, I see that `Upsample`, and `qr`  uses `"mode"` and this actually makes me think we shouldn't use `"mode"` because in both of those cases "mode" indicates *what* is computed as opposed to *how* it is computed. For that reason, I think  `"method"` (which lobpcg uses) or `"algorithm"`/`"strategy"` could be a better choice. I'm sure @albanD has some ideas on this too when he comes back.

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

File: torch/autograd/functional.py [hunk 10]
```diff
@@ -408,7 +421,64 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
...
+        is_outputs_tuple, outputs = output_info
+        # Step 3: for each of the output tangents, split along dim 0
+        jacobian_input_output = []
+        for jac, output_i in zip(outputs_before_split, outputs):
+            jacobian_output_i_output = []
+            for jac, input_j in zip(jac.split(input_numels, dim=0), inputs):
+                # We need to transpose the Jacobian because in forward AD, the
+                # batch dimension represents that of the inputs
+                jacobian_input_i_output_j = jac.permute(*range(1, jac.ndim), 0) \
+                    .reshape(tuple([*output_i.shape, *input_j.shape]))  # noqa: C409
```
Comment (zou3519) [inline torch/autograd/functional.py:469]: This is just `jac.movedim(0, -1).unflatten(-1, *input_j.shape)`

**Review by zou3519 — APPROVED** [review]
code looks correct; might want to wait on Alban for bikeshedding

File: test/test_autograd.py [hunk 11]
```diff
@@ -6607,7 +6616,7 @@ def f(x, y):
 
         # output unrelated to all inputs
         def f(x, y):
-            return torch.randn([])
+            return torch.ones([])
```
Comment (albanD) [inline test/test_autograd.py:6619]: Why this change?
Or is it just that randomness is never working with vmap? Would a `randn_like(x)` be better?

File: torch/autograd/functional.py [hunk 12]
```diff
@@ -1,11 +1,24 @@
 import torch
 from typing import Tuple, List
+from . import forward_ad as fwAD
+from torch._vmap_internals import _vmap  # can we replace with use of is_grads_batched=True
```
Comment (albanD) [inline torch/autograd/functional.py:4]: Not sure wht the comment means?

File: torch/autograd/functional.py [hunk 13]
```diff
@@ -472,6 +547,14 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
          tensor([[3., 0.],
                  [0., 3.]]))
     """
+    assert strategy in ("forward-mode", "reverse-mode")
```
Comment (albanD) [inline torch/autograd/functional.py:550]: There should be a better error message here

File: torch/autograd/functional.py [hunk 14]
```diff
@@ -675,6 +765,7 @@ def hessian(func, inputs, create_graph=False, strict=False, vectorize=False):
     """
 
     is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "hessian")
+    assert outer_jacobian_strategy in ("forward-mode", "reverse-mode")
```
Comment (albanD) [inline torch/autograd/functional.py:768]: same error message

File: torch/autograd/functional.py [hunk 15]
```diff
@@ -433,6 +503,10 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
             because this feature is still experimental, there may be performance
             cliffs. See :func:`torch.autograd.grad`'s ``batched_grad`` parameter for
             more information.
+        strategy (str, optional): Set to ``"forward-mode"`` or ``"reverse-mode"`` to
```
Comment (albanD) [inline torch/autograd/functional.py:506]: I think I am happy with both strategy or mode.

One question though is if forward/reverse the right toggle here. Isn't a toggle that talks about the number of input/outputs of the function easier to understand by the end user?

File: test/test_autograd.py (hunk 11 above)
Comment (soulitzer) [reply test/test_autograd.py:6619]: Yep, it is because randomness does not work with vmap

File: torch/autograd/functional.py (hunk 12 above)
Comment (soulitzer) [reply torch/autograd/functional.py:4]: Good catch, not sure why that is still here. It was from an earlier PR in the stack. I'm removing it.

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).
